// @boost/common@4.1.0/cjs/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const CommonError = require('./CommonError.cjs');
const constants = require('./constants.cjs');
const Contract = require('./Contract.cjs');
const ExitError = require('./ExitError.cjs');
require('./helpers/index.cjs');
const ModulePath = require('./ModulePath.cjs');
const PackageGraph = require('./PackageGraph.cjs');
const Path = require('./Path.cjs');
const PathResolver = require('./PathResolver.cjs');
const Project = require('./Project.cjs');
const json$1 = require('./serializers/json.cjs');
const yaml$1 = require('./serializers/yaml.cjs');
require('./types.cjs');
const VirtualPath = require('./VirtualPath.cjs');
const decorators = require('@boost/decorators');
const createBlueprint = require('./helpers/createBlueprint.cjs');
const deepFreeze = require('./helpers/deepFreeze.cjs');
const deepMerge = require('./helpers/deepMerge.cjs');
const formatMs = require('./helpers/formatMs.cjs');
const instanceOf = require('./helpers/instanceOf.cjs');
const isEmpty = require('./helpers/isEmpty.cjs');
const isFilePath = require('./helpers/isFilePath.cjs');
const isModuleName = require('./helpers/isModuleName.cjs');
const isObject = require('./helpers/isObject.cjs');
const isPlainObject = require('./helpers/isPlainObject.cjs');
const toArray = require('./helpers/toArray.cjs');

/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

exports.CommonError = CommonError.CommonError;
exports.MODULE_NAME_PART = constants.MODULE_NAME_PART;
exports.MODULE_NAME_PATTERN = constants.MODULE_NAME_PATTERN;
exports.Contract = Contract.Contract;
exports.ExitError = ExitError.ExitError;
exports.ModulePath = ModulePath.ModulePath;
exports.PackageGraph = PackageGraph.PackageGraph;
exports.Path = Path.Path;
exports.PathResolver = PathResolver.PathResolver;
exports.Project = Project.Project;
exports.json = json$1;
exports.yaml = yaml$1;
exports.VirtualPath = VirtualPath.VirtualPath;
exports.createBlueprint = createBlueprint.createBlueprint;
exports.deepFreeze = deepFreeze.deepFreeze;
exports.deepMerge = deepMerge.deepMerge;
exports.formatMs = formatMs.formatMs;
exports.instanceOf = instanceOf.instanceOf;
exports.isEmpty = isEmpty.isEmpty;
exports.isFilePath = isFilePath.isFilePath;
exports.isModuleName = isModuleName.isModuleName;
exports.isObject = isObject.isObject;
exports.isPlainObject = isPlainObject.isPlainObject;
exports.toArray = toArray.toArray;
for (const k in decorators) {
  if (k !== 'default' && !exports.hasOwnProperty(k)) Object.defineProperty(exports, k, {
    enumerable: true,
    get: () => decorators[k]
  });
}
