// @boost/terminal@4.0.1/cjs/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const chalk = require('chalk');
const figures = require('figures');
const cursor$1 = require('./cursor.cjs');
const screen$1 = require('./screen.cjs');
const text = require('./text.cjs');
const _interopDefault = e => e && e.__esModule ? e : {
  default: e
};
const chalk__default = /*#__PURE__*/_interopDefault(chalk);
const figures__default = /*#__PURE__*/_interopDefault(figures);

/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

Object.defineProperty(exports, 'style', {
  enumerable: true,
  get: () => chalk__default.default
});
Object.defineProperty(exports, 'figures', {
  enumerable: true,
  get: () => figures__default.default
});
exports.cursor = cursor$1;
exports.screen = screen$1;
exports.ANSI_REGEX = text.ANSI_REGEX;
exports.annotate = text.annotate;
exports.calculateWidth = text.calculateWidth;
exports.hasAnsi = text.hasAnsi;
exports.link = text.link;
exports.sliceAnsi = text.sliceAnsi;
exports.stripAnsi = text.stripAnsi;
exports.truncate = text.truncate;
exports.wrapAnsi = text.wrapAnsi;
