// @boost/log@4.0.1/cjs/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const constants = require('./constants.cjs');
const createLogger = require('./createLogger.cjs');
const formats$1 = require('./formats.cjs');
const Logger = require('./Logger.cjs');
const Transport = require('./Transport.cjs');
const ConsoleTransport = require('./transports/ConsoleTransport.cjs');
const FileTransport = require('./transports/FileTransport.cjs');
const RotatingFileTransport = require('./transports/RotatingFileTransport.cjs');
const StreamTransport = require('./transports/StreamTransport.cjs');
require('./types.cjs');

/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

exports.DEFAULT_LABELS = constants.DEFAULT_LABELS;
exports.LOG_LEVELS = constants.LOG_LEVELS;
exports.MAX_LOG_SIZE = constants.MAX_LOG_SIZE;
exports.createLogger = createLogger.createLogger;
exports.formats = formats$1;
exports.Logger = Logger.Logger;
exports.Transport = Transport.Transport;
exports.ConsoleTransport = ConsoleTransport.ConsoleTransport;
exports.FileTransport = FileTransport.FileTransport;
exports.RotatingFileTransport = RotatingFileTransport.RotatingFileTransport;
exports.StreamTransport = StreamTransport.StreamTransport;
