// @boost/cli@4.0.1/cjs/components/Help.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const React = require('react');
const ink = require('ink');
const constants = require('../constants.cjs');
require('../helpers/index.cjs');
const Header = require('./Header.cjs');
const HelpCommands = require('./internal/HelpCommands.cjs');
const HelpOptions = require('./internal/HelpOptions.cjs');
const HelpParams = require('./internal/HelpParams.cjs');
const HelpUsage = require('./internal/HelpUsage.cjs');
const formatDescription = require('../helpers/formatDescription.cjs');
const _interopDefault = e => e && e.__esModule ? e : {
  default: e
};
const React__default = /*#__PURE__*/_interopDefault(React);

// function extractDefault<K extends string, P>(
// 	key: K,
// ): (result: Record<K, React.ComponentType<P>>) => { default: React.ComponentType<P> } {
// 	return ({ [key]: component }) => ({ default: component });
// }

// const Commands = React.lazy(() =>
// 	import('./internal/HelpCommands').then(extractDefault('HelpCommands')),
// );

// const Options = React.lazy(() =>
// 	import('./internal/HelpOptions').then(extractDefault('HelpOptions')),
// );

// const Params = React.lazy(() => import('./internal/HelpParams').then(extractDefault('HelpParams')));

// const Usage = React.lazy(() => import('./internal/HelpUsage').then(extractDefault('HelpUsage')));
/**
 * A React component that renders a help menu for a command.
 * Includes all sub-commands, options, params, and more.
 */ // eslint-disable-next-line complexity
function Help(props) {
  const categories = props.categories,
    commands = props.commands,
    delimiter = props.delimiter,
    options = props.options,
    header = props.header,
    params = props.params,
    config = props.config;
  const hasDesc = Boolean(config?.description);
  const hasUsage = Boolean(config?.usage && config.usage.length > 0);
  const hasParams = Boolean(params && params.length > 0);
  const hasCommands = Boolean(commands && Object.keys(commands).length > 0);
  const hasOptions = Boolean(options && Object.keys(options).length > 0);
  return /*#__PURE__*/React__default.default.createElement(ink.Box, {
    flexDirection: "column"
  }, (hasDesc || hasUsage) && header && /*#__PURE__*/React__default.default.createElement(Header.Header, {
    label: header
  }), hasDesc && /*#__PURE__*/React__default.default.createElement(ink.Box, {
    paddingLeft: constants.SPACING_COL
  }, /*#__PURE__*/React__default.default.createElement(ink.Text, null, formatDescription.formatDescription(config))), hasUsage && /*#__PURE__*/React__default.default.createElement(HelpUsage.HelpUsage, {
    delimiter: delimiter,
    usage: config?.usage
  }), hasParams && /*#__PURE__*/React__default.default.createElement(HelpParams.HelpParams, {
    config: config,
    params: params
  }), hasCommands && /*#__PURE__*/React__default.default.createElement(HelpCommands.HelpCommands, {
    categories: categories,
    commands: commands
  }), hasOptions && /*#__PURE__*/React__default.default.createElement(HelpOptions.HelpOptions, {
    categories: categories,
    options: options
  }));
}
exports.Help = Help;
