// @boost/config@4.0.1/cjs/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const ConfigError = require('./ConfigError.cjs');
const Configuration = require('./Configuration.cjs');
const getEnv = require('./helpers/getEnv.cjs');
const mergeArray = require('./helpers/mergeArray.cjs');
const mergeExtends = require('./helpers/mergeExtends.cjs');
const mergeObject = require('./helpers/mergeObject.cjs');
const mergePlugins = require('./helpers/mergePlugins.cjs');
const overwrite = require('./helpers/overwrite.cjs');
const schemas = require('./schemas.cjs');
require('./types.cjs');

/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

exports.ConfigError = ConfigError.ConfigError;
exports.Configuration = Configuration.Configuration;
exports.getEnv = getEnv.getEnv;
exports.mergeArray = mergeArray.mergeArray;
exports.mergeExtends = mergeExtends.mergeExtends;
exports.mergeObject = mergeObject.mergeObject;
exports.mergePlugins = mergePlugins.mergePlugins;
exports.overwrite = overwrite.overwrite;
exports.createExtendsSchema = schemas.createExtendsSchema;
exports.createOverridesSchema = schemas.createOverridesSchema;
exports.createPluginsSchema = schemas.createPluginsSchema;
