// @boost/args@4.0.1/cjs/index.cjs
'use strict';

Object.defineProperty(exports, '__esModule', {
  value: true
});
const ArgsError = require('./ArgsError.cjs');
const constants = require('./constants.cjs');
const format = require('./format.cjs');
const parse = require('./parse.cjs');
const ParseError = require('./ParseError.cjs');
const parseInContext = require('./parseInContext.cjs');
require('./types.cjs');
const ValidationError = require('./ValidationError.cjs');

/**
 * @copyright   2020, Miles Johnson
 * @license     https://opensource.org/licenses/MIT
 */

exports.ArgsError = ArgsError.ArgsError;
exports.COMMAND_FORMAT = constants.COMMAND_FORMAT;
exports.DEFAULT_BOOLEAN_VALUE = constants.DEFAULT_BOOLEAN_VALUE;
exports.DEFAULT_NUMBER_VALUE = constants.DEFAULT_NUMBER_VALUE;
exports.DEFAULT_STRING_VALUE = constants.DEFAULT_STRING_VALUE;
exports.LONG_OPTION_FORMAT = constants.LONG_OPTION_FORMAT;
exports.OPTION_LIKE = constants.OPTION_LIKE;
exports.SHORT_OPTION_FORMAT = constants.SHORT_OPTION_FORMAT;
exports.SHORT_OPTION_GROUP_FORMAT = constants.SHORT_OPTION_GROUP_FORMAT;
exports.format = format.format;
exports.parse = parse.parse;
exports.ParseError = ParseError.ParseError;
exports.parseInContext = parseInContext.parseInContext;
exports.ValidationError = ValidationError.ValidationError;
