// @boost/log@4.0.1/cjs/transports/RotatingFileTransport.cjs
'use strict';

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }
function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { _defineProperty(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }
function _defineProperty(obj, key, value) { key = _toPropertyKey(key); if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }
function _toPropertyKey(arg) { var key = _toPrimitive(arg, "string"); return typeof key === "symbol" ? key : String(key); }
function _toPrimitive(input, hint) { if (typeof input !== "object" || input === null) return input; var prim = input[Symbol.toPrimitive]; if (prim !== undefined) { var res = prim.call(input, hint || "default"); if (typeof res !== "object") return res; throw new TypeError("@@toPrimitive must return a primitive value."); } return (hint === "string" ? String : Number)(input); }
Object.defineProperty(exports, '__esModule', {
  value: true
});
const FileTransport = require('./FileTransport.cjs');
const DAYS_IN_WEEK = 7;
class RotatingFileTransport extends FileTransport.FileTransport {
  constructor(...args) {
    super(...args);
    this.lastTimestamp = this.formatTimestamp(Date.now());
  }
  blueprint(schemas) {
    const string = schemas.string;
    return _objectSpread(_objectSpread({}, super.blueprint(schemas)), {}, {
      rotation: string().oneOf(['hourly', 'daily', 'weekly', 'monthly'])
    });
  }

  /**
   * Format a `Date` object into a format used within the log file name.
   */
  formatTimestamp(ms) {
    const rotation = this.options.rotation;
    const date = new Date(ms);
    let timestamp = `${date.getFullYear()}${String(date.getMonth() + 1).padStart(2, '0')}`;
    if (rotation === 'monthly') {
      return timestamp;
    }

    // Special case, calculate the week manually and return,
    // but do not append so other rotations inherit!
    if (rotation === 'weekly') {
      const firstDay = new Date(date.getFullYear(), date.getMonth(), 1).getDay();
      const offsetDate = date.getDate() + firstDay - 1;
      timestamp += `.W${Math.floor(offsetDate / DAYS_IN_WEEK) + 1}`;
      return timestamp;
    }
    timestamp += String(date.getDate()).padStart(2, '0');
    if (rotation === 'daily') {
      return timestamp;
    }
    timestamp += `.${String(date.getHours()).padStart(2, '0')}`;
    return timestamp;
  }

  /**
   * @inheritDoc
   */
  checkIfNeedsRotation() {
    if (this.lastSize > this.options.maxSize || this.formatTimestamp(Date.now()) !== this.lastTimestamp) {
      this.closeStreamAndRotateFile();
    }
  }

  /**
   * @inheritDoc
   */
  getRotatedFileName() {
    const name = this.path.name(true);
    const ext = this.path.ext(true);
    return `${name}-${this.lastTimestamp}.${ext}`;
  }

  /**
   * @inheritDoc
   */
  rotateFile() {
    super.rotateFile();

    // Update timestamp to the new format
    this.lastTimestamp = this.formatTimestamp(Date.now());
  }
}
exports.RotatingFileTransport = RotatingFileTransport;
